packages/next/src/build/analyze/index.ts

```diff
@@ -0,0 +1,216 @@
+import { turbopackAnalyze, type AnalyzeContext } from '../turbopack-analyze'
+import { durationToString } from '../duration-to-string'
+import { cp, writeFile } from 'node:fs/promises'
+import {
+  collectAppFiles,
+  collectPagesFiles,
+  createPagesMapping,
+} from '../entries'
+import { createValidFileMatcher } from '../../server/lib/find-page-file'
+import { findPagesDir } from '../../lib/find-pages-dir'
+import { PAGE_TYPES } from '../../lib/page-types'
+import loadCustomRoutes from '../../lib/load-custom-routes'
+import { generateRoutesManifest } from '../generate-routes-manifest'
+import { checkIsAppPPREnabled } from '../../server/lib/experimental/ppr'
+import { normalizeAppPath } from '../../shared/lib/router/utils/app-paths'
+import http from 'node:http'
+
+// @ts-expect-error types are in @types/serve-handler
+import serveHandler from 'next/dist/compiled/serve-handler'
+import { Telemetry } from '../../telemetry/storage'
+import { eventAnalyzeCompleted } from '../../telemetry/events'
+import { traceGlobals } from '../../trace/shared'
+
+const ANALYZE_PATH = '.next/diagnostics/analyze'
+
+export default async function analyze(
+  dir: string,
+  reactProductionProfiling = false,
+  noMangling = false,
+  appDirOnly = false,
+  serve: boolean,
+  port: number
+): Promise<void> {
+  try {
+    const config: NextConfigComplete = await loadConfig(PHASE_ANALYZE, dir, {
+      silent: false,
+      reactProductionProfiling,
+    })
+
+    process.env.NEXT_DEPLOYMENT_ID = config.deploymentId || ''
+
+    const distDir = path.join(dir, '.next')
+    const telemetry = new Telemetry({ distDir })
+    setGlobal('phase', PHASE_ANALYZE)
+    setGlobal('distDir', distDir)
+    setGlobal('telemetry', telemetry)
+
+    Log.info('Analyzing a production build...')
+
+    const analyzeContext: AnalyzeContext = {
+      config,
+      dir,
+      distDir,
+      noMangling,
+      appDirOnly,
+    }
+
+    let shutdownPromise = Promise.resolve()
+    const { duration: analyzeDuration, shutdownPromise: p } =
+      await turbopackAnalyze(analyzeContext)
+    shutdownPromise = p
+
+    const durationString = durationToString(analyzeDuration)
+    Log.event(`Compiled successfully in ${durationString}`)
+
+    await shutdownPromise
+
+    await cp(
+      path.join(__dirname, '../../bundle-analyzer'),
+      path.join(dir, ANALYZE_PATH),
+      { recursive: true }
+    )
+
+    // Collect and write routes for the bundle analyzer
+    const routes = await collectRoutesForAnalyze(dir, config, appDirOnly)
+
+    // Write an index of routes for the route picker
+    await writeFile(
+      path.join(dir, ANALYZE_PATH, 'data/routes.json'),
+      JSON.stringify(routes, null, 2)
+    )
+
+    telemetry.record(
+      eventAnalyzeCompleted({
+        success: true,
+        durationInSeconds: Math.round(analyzeDuration),
+        totalPageCount: routes.length,
+      })
+    )
+
+    if (serve) {
+      await startServer(ANALYZE_PATH, port)
```

Comment:
Maybe I'm misreading this...

`startServer` resolves the promise as soon as the listen callback succeeds. Does this mean that we'll start the server, but then immediately exit the process?

I think we need to wait until the process gets a SIGINT. Practically that probably means returning a promise that never resolves, because we can rely on the default SIGINT handler to exit the process for us.